src/pint/toa.py

```diff
@@ -2420,11 +2453,66 @@ def merge_TOAs(TOAs_list):
     planets = [tt.planets for tt in TOAs_list]
     if len(set(planets)) > 1:
         raise TypeError(f"merge_TOAs() cannot merge. Inconsistent planets: {planets}")
-    num_cols = [len(tt.table.columns) for tt in TOAs_list]
-    if len(set(num_cols)) > 1:
-        raise TypeError(
-            f"merge_TOAs() cannot merge. Inconsistent numbers of table columns: {num_cols}"
-        )
+    # check for the presence of various computable columns
+    #   pulse_number
+    #   ssb_obs_pos, ssb_obs_vel, obs_sun_pos
+    #   tdb, tdbld
+    # if one file has these and the others don't, try to add them if strict=False
+    all_colnames = [tt.table.colnames for tt in TOAs_list]
+    has_pulse_number = np.array(
+        ["pulse_number" in colnames for colnames in all_colnames]
+    )
+    has_posvel = np.array(
+        [
+            ("ssb_obs_pos" in colnames)
+            and ("ssb_obs_vel" in colnames)
+            and ("obs_sun_pos" in colnames)
+            for colnames in all_colnames
+        ]
+    )
+    has_tdb = np.array(
+        [("tdb" in colnames) and ("tdbld" in colnames) for colnames in all_colnames]
+    )
+    has_posvel_ecl = np.array(
+        ["ssb_obs_vel_ecl" in colnames for colnames in all_colnames]
+    )
+    if has_pulse_number.any() and not has_pulse_number.all():
+        if strict:
+            log.warning("Not all data have 'pulse_number' columns but strict=True")
+        else:
+            # some data have pulse_numbers but not all
+            # put in NaN
+            for i, tt in enumerate(TOAs_list):
+                if not "pulse_number" in tt.table.colnames:
+                    log.warning(
+                        f"'pulse_number' not present in data set {i}: inserting NaNs"
+                    )
+                    tt.table.add_column(np.ones(len(tt)) * np.nan, name="pulse_number")
```

Comment:
Doesn't this modify the input table? I don't think `merge_TOAs` should ever modify its inputs.